### PR TITLE
[AMDGPU][LowerModuleLDS] Refactor partially lowered module detection

### PR DESCRIPTION
**Pull request by Pierre-vh**
Refactor the logic that checks if a module contains mixed absolute/non-lowered LDS GVs.

The check now happens latter when the "worklists" are formed. This is because in some cases (OpenMP) we can have non-lowered GVs in a lowered module, and this is normal because those GVs are just unused and removed from the list at some point before the end of `getUsesOfLDSByFunction`.

Doing the check later ensures that if a mixed module is spotted, then it's a _real_ mixed module that needs rejection, not a module containing an intentionally ignored GV.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp [hunk 1]
```diff
@@ -1143,6 +1153,7 @@ class AMDGPULowerModuleLDS {
   }
 
   bool runOnModule(Module &M) {
+    dbgs() << "LowerModuleLDS: " << M.getName() << " (" << (void *)&M << ")\n";
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:1156]: Leftover debug printing

**Review by arsenm — COMMENTED** [review]
Add test for the unused case?

**Review by arsenm — APPROVED** [review]

**Comment by JonChesterfield** [comment]
I think the control flow around deleting dead variables is a little off. For example, the openmp cause here is a variable in llvm.compiler.used and otherwise dead - removing that despite being in compiler.user seems legitimate as the compiler does nothing with dead lds variables.

An argument could be made for deleting all unused lds variables, ignoring llvm.used as well, since they're module-local and otherwise inaccessible, and I think that is partially reified in the pass. However in the new might-run-multiple-times world, it's not totally clear to me whether an unused variable with an absolute address assigned is dead or not.

We're on slightly thin ice here. The pass was written assuming it was run once on the whole code object. We'd like it to run incrementally and/or somehow support separate linking. That's quite complicated and not an immediate req so we're trying to perturb it to work in the thinlto style case. I can see bugs coming in.

@arsenm did we ever get a reasonable place in tree for runtime tests? As in a bunch of programs written in C++ (ideally) or HIP/opencl (if one must) for things like this? There's some tests in libc which do the right sort of thing but I'm not totally sure I can pitch backend specific execution tests as being a part of libc

**Review by JonChesterfield — APPROVED** [review]
LG, preferably with the extra &

**Comment by arsenm** [comment]
> @arsenm did we ever get a reasonable place in tree for runtime tests? As in a bunch of programs written in C++ (ideally) or HIP/opencl (if one must) for things like this?

This is permanently stuck in "coming soon". There seems to be a fragment of HIP support in llvm-test-suite, but last I tried to use it I couldn't get it to build. It seemed to have odd expectations for how to find rocm

**Pull request merged**